Design doc for distributed autograd.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29175 Design doc for distributed autograd.**

Updates our docs to include a design doc for distributed autograd.
Currently, this doc only covers the FAST mode algorithm. The Smart mode
algorithm section just refers to the original RFC.

There is a section for Distributed Optimizer that we can complete once we've
finalized the API for the same.

Differential Revision: [D18318949](https://our.internmc.facebook.com/intern/diff/D18318949/)

Commit:
Design doc for distributed autograd.

Updates our docs to include a design doc for distributed autograd.
Currently, this doc only covers the FAST mode algorithm. The Smart mode
algorithm section just refers to the original RFC.

There is a section for Distributed Optimizer that we can complete once we've
finalized the API for the same.

Differential Revision: [D18318949](https://our.internmc.facebook.com/intern/diff/D18318949/)

[ghstack-poisoned]

pritamdamania87 commented on docs/source/notes/distributed_autograd.rst:
```diff
@@ -0,0 +1,277 @@
+As an example, this is what the autograd graph for our example above would look 
+like (t5.sum() excluded for simplicity):
+
+.. image:: images/send_recv_functions.png
```
This might be an issue with ghexport, if you look at the phabricator diff the images are present.

jlin27 replied:
Will this be the final folder for the images or will they be moved to https://github.com/pytorch/pytorch/tree/master/docs/source/_static/img

pritamdamania87 replied:
@jlin27 I'm not sure what the convention here is? That folder does seem to have the  logos etc. Do we intend to have design doc specific images there? I feel you feel that folder is a better place, I'll move the images there.

jlin27 replied:
We can add a folder under .../_static/img/autograd similar to how tensorboard (.../_static/imag/tensorboard) did and as long as we reference the direct path, it will get built out in the html.

Reference image from rst in tensorboard:
https://github.com/pytorch/pytorch/blob/master/torch/utils/tensorboard/writer.py#L293

pritamdamania87 commented on docs/source/notes/distributed_autograd.rst:
```diff
@@ -0,0 +1,277 @@
+
+  import torch
+  import torch.distributed.rpc as rpc
+
```
I'll add in `init_rpc` once we have that API ready.

Commit:
Update on "Design doc for distributed autograd."


Updates our docs to include a design doc for distributed autograd.
Currently, this doc only covers the FAST mode algorithm. The Smart mode
algorithm section just refers to the original RFC.

There is a section for Distributed Optimizer that we can complete once we've
finalized the API for the same.

Differential Revision: [D18318949](https://our.internmc.facebook.com/intern/diff/D18318949/)

[ghstack-poisoned]

pritamdamania87 commented:
@mrshenli @jlin27 Could you take another look at this? I do plan to clean up some parts of this doc as other API docs land.

zzzwen commented on docs/source/notes/distributed_autograd.rst:
```diff
@@ -0,0 +1,278 @@
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Each forward and backward pass that uses distributed autograd is assigned a 
+unique ``Distributed Autograd Context`` and this context has a globally unique 
```
It will be really great if this can be a clickable link to the source code.

pritamdamania87 replied:
We're working on docs: https://github.com/pytorch/pytorch/pull/29276 and can update this document to link to that when it is ready.

mrshenli commented on docs/source/notes/distributed_autograd.rst:
```diff
@@ -0,0 +1,278 @@
+.. code::
+
+  import torch.distributed.autograd as dist_autograd
+  with dist_autograd.context() as context_id:
```
do we need `as context_id` here?

mrshenli reviewed (approved):
Contents LGTM! Thanks! Please get a stamp from @jlin27 as well.

jlin27 commented on docs/source/notes/distributed_autograd.rst:
```diff
@@ -0,0 +1,278 @@
+we attach ``send`` and ``recv`` functions to the autograd graph when we perform 
+an RPC.
+
+- The ``send`` function is attached to the source of the RPC and it's output 
```
Change "it's" --> "its"

jlin27 commented on docs/source/notes/distributed_autograd.rst:
```diff
@@ -0,0 +1,278 @@
+
+- The ``send`` function is attached to the source of the RPC and it's output 
+  edges point to the autograd function for the input tensors of the RPC. 
+  The input for this function is received over the wire from the appropriate 
```
Suggestion : The phrase "received over the wire" is very American idiomatic. Can be challenging for non-native speakers to understand/difficult to translate. Consider simplifying to just "is received from the appropriate..." or something similar.

jlin27 commented on docs/source/notes/distributed_autograd.rst:
```diff
@@ -0,0 +1,278 @@
+  ``recv`` function.
+- The ``recv`` function is attached to the destination of the RPC and its 
+  inputs are retrieved from operators executed on the destination using the 
+  input tensors. The output gradients of this function are passed over the wire 
```
Suggestion : Similar comment about the phrase "received over the wire" as above.

jlin27 commented on docs/source/notes/distributed_autograd.rst:
```diff
@@ -0,0 +1,278 @@
+
+The distributed autograd graph with dependencies would be as follows:
+
+.. image:: images/distributed_dependencies_computed.png
```
Callout to replace this with the final image path

jlin27 commented on docs/source/notes/distributed_autograd.rst:
```diff
@@ -0,0 +1,278 @@
+4. After computing dependencies, kick off the local autograd engine with the 
+   provided roots.
+5. When the autograd engine executes the ``recv`` function, the ``recv`` 
+   function sends the input gradients over the wire to the appropriate worker. 
```
Suggestion to remove the idiomatic phrase "over the wire" that's similar to comment above.

jlin27 commented on docs/source/notes/distributed_autograd.rst:
```diff
@@ -0,0 +1,278 @@
+   `Distributed Autograd Context`_. The gradients are stored in a 
+   ``Dict[Tensor, Tensor]``, which is basically a map from Tensor to its 
+   associated gradient and this map can be retrieved using the 
+   ``get_gradients`` API.
```
Suggestion: Given how many moving parts there are in this process, consider adding a diagram of this flow to visual readers.

jlin27 replied:
Got it

Commit:
Update on "Design doc for distributed autograd."


Updates our docs to include a design doc for distributed autograd.
Currently, this doc only covers the FAST mode algorithm. The Smart mode
algorithm section just refers to the original RFC.

There is a section for Distributed Optimizer that we can complete once we've
finalized the API for the same.

Differential Revision: [D18318949](https://our.internmc.facebook.com/intern/diff/D18318949/)

[ghstack-poisoned]
